### PR TITLE
ceph: add missing python library

### PR DESCRIPTION
**Pull request by dermetfan**
This is a regression in 22.11. After updating from 22.05 the dashboard no longer works and `ceph status` reports a health check error.

The `routes` python library required by the mgr dashboard module is missing.

`ceph -s` reports:

```
  cluster:
    id:     …
    health: HEALTH_ERR
            Module 'dashboard' has failed: No module named 'routes'
```

`ceph -w` reports:

```
mgr.b [ERR] Unhandled exception from module 'dashboard' while running on mgr.b: No module named 'routes'
mon.b [ERR] Health check failed: Module 'dashboard' has failed: No module named 'routes' (MGR_MODULE_ERROR)
```

###### Description of changes

This simply adds the `routes` library to the python environment.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - ~~and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)~~
  - ~~or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)~~
  - ~~made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages~~
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- ~~[23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)~~
  - [ ] ~~(Package updates) Added a release notes entry if the change is major or breaking~~
  - [ ] ~~(Module updates) Added a release notes entry if the change is significant~~
  - [ ] ~~(Module addition) Added a release notes entry if adding a new NixOS module~~
  - [ ] ~~(Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes~~
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
Can you add the error you saw in `ceph -s` as well as how you set up the dashboard? It would be good to know, so we can add a test to keep this working in the future.

I would also be interested to know where ceph defines those python dependencies. Ideally we'd be looking them up, but I can't find `routes` anywhere in the 16.2.10 tarball.

**Comment by dermetfan** [comment]
> Can you add the error you saw in `ceph -s` […]?

Sure, updated the description.

> Can you add […] how you set up the dashboard?

I really only ran these commands via [this](https://git.sr.ht/~dermetfan/garnix/tree/2110434167bb63453061752bd4a68452c7e652b0/item/parts/nixosModules/services/ceph/mgr-dashboard.nix#L35,37) NixOS module:

```
ceph mgr module enable dashboard
ceph config set mgr mgr/dashboard/b/server_port 6780
```

**Comment by dermetfan** [comment]
Looks like the python dependencies are only listed in the [debian package](https://github.com/ceph/ceph/blob/6fec0c82d7bc3961943db56eb945b651f567b204/debian/ceph-mgr-dashboard.requires#L7).

File: pkgs/tools/filesystems/ceph/default.nix [hunk 1]
```diff
@@ -115,6 +115,7 @@ let
   ceph-python-env = python.withPackages (ps: [
```
Comment (mweinelt) [inline pkgs/tools/filesystems/ceph/default.nix:115]:
```suggestion
  ceph-python-env = python.withPackages (ps: [
    # Check .requires files below https://github.com/ceph/ceph/tree/main/debian for dependencies 
```

**Comment by mweinelt** [comment]
Tried to make a test out of this, the cluster remains healthy, but the dashboard does not start listening on the given port.

```
diff --git a/nixos/tests/ceph-multi-node.nix b/nixos/tests/ceph-multi-node.nix
index 556546beee7..b8e17bee211 100644
--- a/nixos/tests/ceph-multi-node.nix
+++ b/nixos/tests/ceph-multi-node.nix
@@ -215,6 +215,17 @@ let
     monA.wait_until_succeeds("ceph osd stat | grep -e '3 osds: 3 up[^,]*, 3 in'")
     monA.wait_until_succeeds("ceph -s | grep 'mgr: ${cfg.monA.name}(active,'")
     monA.wait_until_succeeds("ceph -s | grep 'HEALTH_OK'")
+
+    # Enable the dashboard and recheck health
+    monA.succeed("ceph mgr module enable dashboard")
+    monA.succeed("ceph config set mgr mgr/dashboard/${cfg.monA.name}/server_port 8080")
+    monA.log(monA.execute("ceph -s")[1])
+    monA.sleep(10)
+    monA.log(monA.execute("ceph -s")[1])
+    monA.wait_for_open_port(8080)
+    monA.log(monA.execute("ss -lpn | grep 8080")[1])
+    monA.wait_until_succeeds("curl -q -fail http://localhost:8080")
+    monA.wait_until_succeeds("ceph -s | grep 'HEALTH_OK'")
   '';
 in {
   name = "basic-multi-node-ceph-cluster";

**Comment by dermetfan** [comment]
Turns out SSL is enabled by default and it waits for a key to be configured. Disabled that and it starts right up.

**Comment by mweinelt** [comment]
Giving it a quick spin w/o the fix to test the negative case.

**Comment by mweinelt** [comment]
@ofborg test ceph-single-node

**Comment by mweinelt** [comment]
Fails successfully due to the missing library. :partying_face: 

```
monA # [  198.672953] ceph-mgr[5931]: 2023-01-20T23:35:50.854+0000 7f0567fff640 -1 dashboard.serve:                                               
monA # [  198.680830] ceph-mgr[5931]: 2023-01-20T23:35:50.854+0000 7f0567fff640 -1 Traceback (most recent call last):
monA # [  198.682508] ceph-mgr[5931]:   File "/nix/store/9dqwapw4b9r4vxgh2dccmc4x255wxhr7-ceph-16.2.10-lib/lib/ceph/mgr/dashboard/module.py", line 335, in serve
monA # [  198.684791] ceph-mgr[5931]:     mapper, parent_urls = Router.generate_routes(self.url_prefix)                                           
monA # [  198.686271] ceph-mgr[5931]:   File "/nix/store/9dqwapw4b9r4vxgh2dccmc4x255wxhr7-ceph-16.2.10-lib/lib/ceph/mgr/dashboard/controllers/_router.py", line 52, in generate_routes
monA # [  198.688732] ceph-mgr[5931]:     mapper = cherrypy.dispatch.RoutesDispatcher()                                                           
monA # [  198.690018] ceph-mgr[5931]:   File "/nix/store/079sjiv8nbahmhlfsw3l1ly9grfi0x6s-python3-3.9.16-env/lib/python3.9/site-packages/cherrypy/_cpdispatch.py", line 506, in __init__
monA # [  198.692549] ceph-mgr[5931]:     import routes                                                                                           
monA # [  198.693451] ceph-mgr[5931]: ModuleNotFoundError: No module named 'routes'                                                               
monA # [  200.339276] ceph-mon[873]: 2023-01-20T23:35:52.518+0000 7f27fd7fb640 -1 log_channel(cluster) log [ERR] : Health check failed: Module 'dashboard' has failed: No module named 'routes' (MGR_MODULE_ERROR)

**Comment by mweinelt** [comment]
Ideally at some point we'd split up the tests based on feature set. But that day is not today.

**Pull request merged**